### PR TITLE
allow adding annotations to pod when using kubectl run

### LINKED ISSUE #826: Allow adding annotations to pod when using kubectl run
When using `kubectl run` to run a shell for debug/diagnostic purposes, it would be really useful to be able to add annotations to the created pod. The use case I have in mind is to control sidecar injectors, for example, if you're using Istio with automatic sidecar injection, you may not want the created pod to have the sidecar injected, to do this you need to pass set the annotation `sidecar.istio.io/inject: false`. Or, if you have Istio configured to opt in to sidecar injection, you may decide that you want your debug shell to have the sidecar injected, in which case you would want to set `sidecar.istio.io/inject: true`.

Here's an example of what using this feature might look like:

```
kubectl run -i --tty --rm centos --image=centos --restart=Never --annotation=sidecar.istio.io/inject=false -- sh
```

### PR DESCRIPTION
**Pull request by SaiHarshaK**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Allows adding annotations to the created pod from kubectl run through --annotation flag.

**Which issue(s) this PR fixes**:
Fixes [kubernetes/kubectl#826](https://github.com/kubernetes/kubectl/issues/826)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/run/run.go [hunk 1]
```diff
@@ -173,6 +173,7 @@ func NewCmdRun(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Co
 
 func addRunFlags(cmd *cobra.Command, opt *RunOptions) {
 	cmdutil.AddDryRunFlag(cmd)
+	cmd.Flags().String("annotation", "", "Comma separated annotations to apply to the pod.")
```
Comment (pwittrock) [inline staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:176]: Why not make this a StringArray?

Comment (SaiHarshaK) [reply staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:176]: I followed the labels flag for it, I'll change it

File: staging/src/k8s.io/kubectl/pkg/cmd/run/run.go [hunk 2]
```diff
@@ -173,6 +173,7 @@ func NewCmdRun(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Co
 
 func addRunFlags(cmd *cobra.Command, opt *RunOptions) {
 	cmdutil.AddDryRunFlag(cmd)
+	cmd.Flags().StringArray("annotation", []string{}, "Annotations to apply to the pod.")
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:176]: Wrap the description in `i18n.T()` like the others. This is used for internationalization/translation.

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/run.go [hunk 3]
```diff
@@ -87,6 +87,25 @@ func getArgs(genericParams map[string]interface{}) ([]string, error) {
 	return args, nil
 }
 
+// getAnnotations returns map of annotations.
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/generate/versioned/run.go:90]: Is there a way you can use [this](https://github.com/kubernetes/kubernetes/blob/b497fa1d34ec15c4219d591e541c2cf811271357/staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go#L380-L453) all here?

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/run.go [hunk 4]
```diff
@@ -327,6 +353,22 @@ func (BasicPod) Generate(genericParams map[string]interface{}) (runtime.Object,
 	return &pod, nil
 }
 
+// parseAnnotations converts []string into map of annotations.
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/generate/versioned/run.go:356]: Is there a way you can use [this](https://github.com/kubernetes/kubernetes/blob/b497fa1d34ec15c4219d591e541c2cf811271357/staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go#L380-L453) all here?

**Review by eddiezane — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/run.go [hunk 5]
```diff
@@ -87,6 +88,25 @@ func getArgs(genericParams map[string]interface{}) ([]string, error) {
 	return args, nil
 }
 
+// getAnnotations returns map of annotations.
+func getAnnotations(genericParams map[string]interface{}) (map[string]string, error) {
+	var annotations map[string]string
+	annotationStrings, found := genericParams["annotation"]
+	if found {
```
[the remark this answers is not in the document]
Comment (SaiHarshaK) [reply staging/src/k8s.io/kubectl/pkg/generate/versioned/run.go:95]: I'll change it. I actually followed getEnv() function for the logic.

File: staging/src/k8s.io/kubectl/pkg/cmd/run/run.go (hunk 2 above)
Comment (SaiHarshaK) [reply staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:176]: Right, my bad. Shall i also add `i18n.T()` to labels in this PR or only wrap it around annotations?

**Review by eddiezane — APPROVED** [review]
Great job on this!

/lgtm
/assign @pwittrock

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/run_test.go [hunk 6]
```diff
@@ -20,7 +20,7 @@ import (
 	"reflect"
 	"testing"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (pwittrock) [inline staging/src/k8s.io/kubectl/pkg/generate/versioned/run_test.go:23]: why is this necessary?

**Comment by pwittrock** [comment]
FYI: This can wait until post 1.19 to merge

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/run_test.go (hunk 6 above)
Comment (SaiHarshaK) [reply staging/src/k8s.io/kubectl/pkg/generate/versioned/run_test.go:23]: vscode added it, but I kept it because I thought maybe golang standard wants us using that specifically when multiple paths have same pkg name.

**Comment by pwittrock** [comment]
Woot.
